pkg/kv/kvserver/allocator/storepool/store_pool.go

```diff
@@ -558,26 +571,50 @@ func (sp *StorePool) statusString(nl NodeLivenessFunc) string {
 // storeGossipUpdate is the Gossip callback used to keep the StorePool up to date.
 func (sp *StorePool) storeGossipUpdate(_ string, content roachpb.Value) {
 	var storeDesc roachpb.StoreDescriptor
-	// We keep copies of the capacity and storeID to pass into the
-	// capacityChanged callback.
-	var oldCapacity, curCapacity roachpb.StoreCapacity
-	var storeID roachpb.StoreID
 
 	if err := content.GetProto(&storeDesc); err != nil {
 		ctx := sp.AnnotateCtx(context.TODO())
 		log.Errorf(ctx, "%v", err)
 		return
 	}
-	storeID = storeDesc.StoreID
-	curCapacity = storeDesc.Capacity
+
+	sp.storeDescriptorUpdate(storeDesc)
+}
+
+// storeDescriptorUpdate takes a store descriptor and updates the corresponding
+// details for the store in the storepool.
+func (sp *StorePool) storeDescriptorUpdate(storeDesc roachpb.StoreDescriptor) {
+	// We keep copies of the capacity and storeID to pass into the
+	// capacityChanged callback.
+	var oldCapacity roachpb.StoreCapacity
+	storeID := storeDesc.StoreID
+	curCapacity := storeDesc.Capacity
+
+	now := sp.clock.PhysicalTime()
 
 	sp.DetailsMu.Lock()
 	detail := sp.GetStoreDetailLocked(storeID)
+
+	// We update the store descriptor to reflect the maximum IO Overload values
+	// that have been seen in the past tracking interval, updating the trackers
+	// with the new values prior to querying. This means that the most recent
+	// value for  IOThreshold.*Threshold is always used, together with the
+	// tracked recent maximum IOThreshold.L0NumFiles/L0NumSubLevels to
+	// calculate the score.
+	detail.storeHealthTracker.NumL0FilesTracker.Record(
+		now, float64(curCapacity.IOThreshold.L0NumFiles))
+	detail.storeHealthTracker.NumL0SublevelsTracker.Record(
+		now, float64(curCapacity.IOThreshold.L0NumSubLevels))
+	maxL0NumFiles, _ := detail.storeHealthTracker.NumL0FilesTracker.Query(now)
+	maxL0NumSublevels, _ := detail.storeHealthTracker.NumL0SublevelsTracker.Query(now)
+	storeDesc.Capacity.IOThreshold.L0NumFiles = int64(maxL0NumFiles)
+	storeDesc.Capacity.IOThreshold.L0NumSubLevels = int64(maxL0NumSublevels)
```

Comment:
Something smells fishy here. Are we stamping on follower pausing by updating this store capacity IO threshold state with a rolling 10m max? We're using this same state for follower pausing here:

https://github.com/cockroachdb/cockroach/blob/2111f9bb16a04eb7f27d9a99ffb5cd870ebfce2c/pkg/kv/kvserver/store_raft.go#L787-L791

Which I take is updated for every store every 15s:

https://github.com/cockroachdb/cockroach/blob/2cac11b0b649a7c52ca600a114c60945f77ac5cd/pkg/util/admission/grant_coordinator.go#L100-L116

Reply:
That is a problem and I didn't check where the follower pausing was using the storepool to get the descriptors. The easiest way to swap in IOThreshold was to use the store descriptor capacity. The store descriptor is taken directly from the storepool. 

I am going to add a new type to encapsulate the descriptor and derived capacity information. Then replace calls to desc.Capacity with the derived capacity everywhere in the allocator and storepool update code.